crates/ghost_actor/src/lib.rs

```diff
@@ -443,20 +443,99 @@ mod tests {
-            for (rid, msg) in self.dht.drain_responses() {
-                let mut cb = std::mem::replace(&mut self.dht_callbacks, None);
-                cb.as_mut()
-                    .expect("exists")
-                    .handle(rid, self.as_any(), msg)?;
-                std::mem::replace(&mut self.dht_callbacks, cb);
+            let mut dht_channel = std::mem::replace(&mut self.dht_channel, None);
+            dht_channel
+                .as_mut()
+                .expect("exists")
+                .process(self.as_any())?;
+            std::mem::replace(&mut self.dht_channel, dht_channel);
+            let mut channel_self = std::mem::replace(&mut self.channel_self, None);
+            channel_self
+                .as_mut()
+                .expect("exists")
+                .process(self.as_any())?;
+            std::mem::replace(&mut self.channel_self, channel_self);
+
+            for mut msg in self.channel_self.as_mut().expect("exists").drain_requests() {
+                match msg.take_payload().expect("exists") {
+                    RequestToChild::Bind { spec } => {
+                        // do some internal bind
+                        // we get a bound_url
+                        let bound_url = "bound_url".to_string();
+                        // respond to our parent
+                        msg.respond(Ok(RequestToChildResponse::Bind(BindResultData {
+                            bound_url: bound_url,
+                        })));
+                    }
+                    RequestToChild::Bootstrap { address } => {}
+                    RequestToChild::SendMessage { address, payload } => {
+                        self.dht_channel.as_mut().expect("exists").request(
+                            std::time::Duration::from_millis(2000),
+                            GwDht::ResolveAddressForId { msg },
+                            dht_protocol::RequestToChild::ResolveAddressForId { id: address },
+                            Box::new(|m, context, response| {
+                                let m = match m.downcast_mut::<GatewayTransport>() {
+                                    None => panic!("wrong type"),
+                                    Some(m) => m,
+                                };
+
+                                let msg = {
+                                    if let GwDht::ResolveAddressForId { msg } = context {
+                                        msg
+                                    } else {
+                                        panic!("bad context type");
+                                    }
+                                };
+
+                                // got a timeout error
+                                if let GhostCallbackData::Timeout = response {
+                                    msg.respond(Err("Timeout".into()));
+                                    return Ok(());
+                                }
+
+                                let response = {
```

Comment:
especially all this response handling, but maybe that's the whole point because this is how we are implementing a Transport level response system, in a Actor level response system... :-)  I guess we're just going to have to comment the heck of this... !